src/typegraph/core/src/global_store.rs

```diff
@@ -157,9 +158,8 @@ impl Store {
         with_store(|s| s.type_by_names.get(name).map(|id| id.0.clone()))
     }
 
-    pub fn register_type_ref(builder: TypeRefBuilder) -> Result<TypeRef> {
-        let id = with_store(|s| s.types.len()) as u32;
-        let type_ref = builder.with_id(id.into());
+    pub fn assign_type_ref(id: TypeId, builder: TypeRefBuilder) -> Result<TypeRef> {
+        let type_ref = builder.with_id(id);
```

Comment:
I am not sure if this can be "safely" used this sort of implies of ability to override. Imo TypeId should not be the responsability of something external to the type constructors.